hybrid/pkg/server/server_interceptor.go

```diff
@@ -0,0 +1,119 @@
+package hybrid_server
+
+import (
+	"context"
+
+	hclog "github.com/hashicorp/go-hclog"
+	nodeattestorv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/server/nodeattestor/v1"
+)
+
+type ServerInterceptorInterface interface {
+	Recv() (*nodeattestorv1.AttestRequest, error)
+	Send(resp *nodeattestorv1.AttestResponse) error
+	setCustomStream(stream nodeattestorv1.NodeAttestor_AttestServer)
+	SetContext(ctx context.Context)
+	Context() context.Context
+	SetLogger(logger hclog.Logger)
+	SetReq(req *nodeattestorv1.AttestRequest)
+	CanReattest() []bool
+	SpiffeID() string
+	SetSpiffeID(spiffeID string)
+	CombinedSelectors() []string
+	Stream() nodeattestorv1.NodeAttestor_AttestServer
+	ResetInterceptor()
+	SpawnInterceptor() ServerInterceptorInterface
+}
+
+type HybridPluginServerInterceptor struct {
+	ctx    context.Context
+	stream nodeattestorv1.NodeAttestor_AttestServer
+	nodeattestorv1.NodeAttestor_AttestServer
+	logger            hclog.Logger
+	req               *nodeattestorv1.AttestRequest
+	Response          *nodeattestorv1.AttestResponse
+	combinedSelectors []string
+	spiffeID          string
+	canReattest       []bool
+}
+
+func (m *HybridPluginServerInterceptor) ResetInterceptor() {
+	m.ctx = nil
+	m.stream = nil
+	m.logger = nil
+	m.req = nil
+	m.Response = nil
+	m.combinedSelectors = nil
+	m.spiffeID = ""
+	m.canReattest = nil
+}
+
+func (m *HybridPluginServerInterceptor) SpawnInterceptor() ServerInterceptorInterface {
+	return &HybridPluginServerInterceptor{
+		ctx:               m.ctx,
+		stream:            m.stream,
+		logger:            m.logger,
+		req:               m.req,
+		Response:          m.Response,
+		combinedSelectors: m.combinedSelectors,
+		spiffeID:          m.spiffeID,
+		canReattest:       m.canReattest,
+	}
+}
+
+func (m *HybridPluginServerInterceptor) Recv() (*nodeattestorv1.AttestRequest, error) {
+	return m.req, nil // add error here
+}
+
+func (m *HybridPluginServerInterceptor) setCustomStream(stream nodeattestorv1.NodeAttestor_AttestServer) {
+	m.stream = stream
+}
+
+func (m *HybridPluginServerInterceptor) Send(resp *nodeattestorv1.AttestResponse) error {
+	if x, ok := resp.GetResponse().(*nodeattestorv1.AttestResponse_AgentAttributes); ok {
```

Comment:
we don't expect that the send method returns anything. It is used by the plugins only when they are done for and closing the stream. In case it reaches the send method it means that all other processing has been done and it's ready to close the stream. Then we capture the output of the send method and store it, That's it. If any errors occurs before the send method gets called, then the plugin will shutdown itself. This behavior is correct and expected :)